File: Python/006_ZigZag_Conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows <= 1:
            return s
            
        slist = [""] * numRows
        i, direction = 0, 1
        for c in s:
            slist[i] += c
            if i == 0 and direction == -1:
                direction = 1
            elif i == numRows-1 and direction == 1:
                direction = -1
            i += direction
        
        return "".join(slist)
    
    
class Solution2(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows == 1:
            return s
        
        strlist = ["" for _ in xrange(numRows)]
        i, direction = 0, 1
        for c in s:
            strlist[i] += c
            i += direction
            if i == 0 or i == numRows-1:
                direction = -direction
        return "".join(strlist)

if __name__ == "__main__":
    arguments = [("PAYPALISHIRING", 3)]
    answers = ["PAHNAPLSIIGYIR"]
    from leetcodelib import test
    test(Solution().convert, arguments, answers)

File: C++/006_ZigZag_Conversion.cpp
class Solution1 {
public:
    string convert(string s, int numRows) {
        if (numRows <= 1) return s;
        string res;
        int k = numRows-1, ind = -1;
        for (int step = k; step >= 0; step--) {
        	for (int i = 0; i <= s.size()/(k*2); i++) {
        		int center = k + k * 2 * i;
        		int inds[2] = {center-step, center+step};
        		for (int j = 0; j < 2; j++) {
        			if (inds[j] != ind && inds[j] < s.size()) {
        				ind = inds[j];
        				res.push_back(s[ind]);
        			}
        		}
        	}
        }
        return res;
    }
};

class Solution2 {
public:
    string convert(string s, int numRows) {
        if (numRows <= 1) return s;
        string *l = new string[numRows];
        int direction = 1, row = 0;
        for (int i = 0; i < s.size(); i++) {
        	l[row].push_back(s[i]);
        	row += direction;
        	if (row == 0 || row == numRows-1) direction = -direction;
        }
        string res;
        for (int i = 0; i < numRows; i++) res.append(l[i]);
        delete[] l;
        return res;
    }
};
